core/lib/zksync_core/src/sync_layer/gossip/conversions.rs

```diff
@@ -0,0 +1,88 @@
+//! Conversion logic between server and consensus types.
+
+use anyhow::Context as _;
+use serde::{Deserialize, Serialize};
+
+use zksync_consensus_roles::validator::{
+    BlockHeader, BlockHeaderHash, BlockNumber, CommitQC, FinalBlock, Payload,
+};
+use zksync_types::block::ConsensusBlockFields;
+use zksync_types::{
+    api::en::SyncBlock, Address, L1BatchNumber, MiniblockNumber, ProtocolVersionId, H256,
+};
+
+use crate::sync_layer::fetcher::FetchedBlock;
+
+// FIXME: should use Protobuf
+#[derive(Debug, Serialize, Deserialize)]
+struct BlockPayload {
+    hash: H256,
+    l1_batch_number: L1BatchNumber,
+    timestamp: u64,
+    l1_gas_price: u64,
+    l2_fair_gas_price: u64,
+    virtual_blocks: u32,
+    operator_address: Address,
+    transactions: Vec<zksync_types::Transaction>,
+}
+
+pub(super) fn sync_block_to_payload(block: SyncBlock) -> Payload {
+    let payload = serde_json::to_vec(&BlockPayload {
+        hash: block.hash.unwrap_or_default(),
+        l1_batch_number: block.l1_batch_number,
+        timestamp: block.timestamp,
+        l1_gas_price: block.l1_gas_price,
+        l2_fair_gas_price: block.l2_fair_gas_price,
+        virtual_blocks: block.virtual_blocks.unwrap_or(0),
+        operator_address: block.operator_address,
+        transactions: block
+            .transactions
+            .expect("Transactions are always requested"),
+    });
+    Payload(payload.expect("Failed serializing block payload"))
+}
+
+pub(super) fn sync_block_to_consensus_block(mut block: SyncBlock) -> anyhow::Result<FinalBlock> {
+    let number = BlockNumber(block.number.0.into());
+    let consensus = block.consensus.take().context("Missing consensus fields")?;
+    let prev_block_hash = consensus.prev_block_hash;
+    let payload = sync_block_to_payload(block);
+    let header = BlockHeader {
+        parent: BlockHeaderHash::from_bytes(prev_block_hash.0),
+        number,
+        payload: payload.hash(),
+    };
+    let justification: CommitQC = zksync_consensus_schema::decode(&consensus.commit_qc_bytes.0)
+        .context("Failed deserializing commit QC from Protobuf")?;
+    Ok(FinalBlock {
+        header,
+        payload,
+        justification,
+    })
+}
+
+impl FetchedBlock {
+    pub(super) fn from_gossip_block(block: &FinalBlock) -> anyhow::Result<Self> {
+        let number = u32::try_from(block.header.number.0)
+            .context("Integer overflow converting block number")?;
+        let payload: BlockPayload = serde_json::from_slice(&block.payload.0)
+            .context("Failed deserializing block payload")?;
+
+        Ok(Self {
+            number: MiniblockNumber(number),
+            l1_batch_number: payload.l1_batch_number,
+            protocol_version: ProtocolVersionId::latest(), // FIXME
```

Comment:
I'm not sure where to get this field from. Consensus messages have their own protocol versions, but I'm not 100% sure whether they should be the same (if so, it needs to be changed on the consensus side and probably shouldn't be made a constant, since protocol versions are semi-frequently updated).

Reply:
The current idea is that zksync-era protocol version is supposed to imply the ProtocolVersion in era-consensus (i.e. we can make it the same, just not all protocol changes will affect the consensus itself). You can extract the protocol version from commit_qc (it contains the ReplicaCommit message which contains the protocol_version field). It will be eventually set by the main node logic, currently it is just a constant.

Reply:
I've created a task for this in Linear. Pragmatically, I'd prefer to leave a FIXME here because otherwise we'd need to mock retrieval of the past protocol versions in tests.